src/lib/reasoners/rel_utils.ml

```diff
@@ -193,126 +193,179 @@ end = struct
+  val subst : Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ex p v cs] replaces all the instances of [p] with [v] in the
+      constraints.
 
-  val remove : X.r -> t -> t
-  (** [remove r t] unregisters the semantic value [r] from the congruence.
+      Use this to ensure that the representation is always normalized.
 
-      Note that if substitutions have been applied to the congruence after a
-      value has been added, those same substitutions must be applied to the
-      semantic value prior to calling [remove], or [Invalid_argument] will be
-      raised.
+      The explanation [ex] justifies the equality [p = v]. *)
 
-      @raise [Invalid_argument] if [r] is not a registered semantic value. *)
+  val add : t -> Constraint.t -> t
+  (** [add c cs] adds the constraint [c] to [cs]. *)
 
-  val subst : X.r -> X.r -> t -> (X.r -> X.r -> 'a -> 'a) -> 'a -> t * 'a
-  (** [subst p v t f x] performs a local congruence closure of the
-      substitution [p -> v].
+  val fold_fresh : (Constraint.t -> 'a -> 'a) -> t -> 'a -> t * 'a
+  (** [fold_fresh f cs acc] folds [f] over the fresh constraints in [cs].
 
-      More precisely, it will fold [f] over the pairs [(rr, nrr)] such that:
-      - [rr] was registered in the congruence
-      - [nrr] is [X.subst p v rr]
+      Fresh constraints are constraints that were never propagated yet. *)
 
-      For each such pair, [rr] is then unregistered from the congruence, and
-      [nrr] is registered instead.
+  val fold_r : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
+  (** [fold_r f cs acc] folds [f] over any representative [r] that is currently
+      associated with a constraint (i.e. at least one constraint currently
+      applies to [r]). *)
 
-      [f] is intended to perform a substitution operation on the type ['a],
-      merging the values associated with [rr] into the values associated with
-      [nrr]. *)
+  val propagate : t -> X.r -> Constraint.domain -> Constraint.domain
+  (** [propagate cs r dom] propagates the constraints associated with [r] in the
+      constraint set [cs] and returns the new domain map after propagation. *)
 end = struct
-  module SX = Shostak.SXH
+  module IM = Util.MI
   module MX = Shostak.MXH
 
-  type t =
-    { parents : SX.t MX.t
-    (** Map from leaves to terms that contain them as a leaf.
-
-        [p] is in [parents(x)] => [x] is in [leaves(p)] *)
-    ; registered : SX.t
-    (** The set of terms we care about. If [x] is in [registered],
-        then [x] is also in [parents(y)] for each [y] in [leaves(x)]. *)
-    }
+  module CS = Set.Make(Constraint)
 
-  let empty = { parents = MX.empty ; registered = SX.empty }
+  type t = {
+    cs_set : CS.t ;
+    (*** All the constraints currently active *)
+    cs_map : CS.t MX.t ;
+    (*** Mapping from semantic values to the constraints that involves them *)
+    fresh : CS.t ;
+    (*** Fresh constraints that have never been propagated *)
```

Comment:
I don't understand the meaning of this field. We add a constraint to it in `add` and we flush it in `fold_fresh` but we don't touch it in `propagate`.

Reply:
Because `propagate` propagates the constraints associated with a single representative `r`. The `fresh` field is cleared by `fold_fresh` which is intended to be used as `Constraints.fold_fresh Constraint.propagate` (see usage in `bitv_rel.ml`). I will refactor to expose `propagate_fresh` instead of `fold_fresh` which makes more sense.